docs/autorenew-feature.md

```diff
@@ -0,0 +1,71 @@
+# The autorenew feature
+
+## Goal
+-	Allow Hedera Services to automatically search for entities such as accounts, files, smart contracts, topics, tokens... that are about to expire and renew them.
+
+## Design
+- Introduce new settings in `application.properties`:
+  * `autorenew.isEnabled`
+  * `autorenew.numberOfEntitiesToCheck`
+  * `autorenew.maxNumberOfEntitiesToRenew`
+  * `autorenew.gracePeriod`
+- Each Hedera entity has an `expirationTime` which is the effective consensus timestamp at (and after) which the entity is set to expire.
+- Each Hedera entity also has an `autoRenewAccount` which is the account to pay for the fee at renewal. This `autoRenewAccount` could be itself if the entity is a crypto account or an account associated with the entity when it was created.
+- When a Hedera entity is created, its initial lifetime is defined by its `autoRenewPeriod`. At its `expirationTime`, Hedera Services will try to extend an entity's lifetime by another `autoRenewPeriod` if the `autoRenewAccount` has enough balance to do so, or as much extension as the remaining balance permits.
+- After handling a transaction, Hedera Services will search within the next `autorenew.numberOfEntitiesToCheck` for upto `autorenew.maxNumberOfEntitiesToRenew` entities that expired then try to renew these entities.
+- After the grace period, if the `expirationTime` of an entity is not extended, it will be deleted from the system.
+
+## Implementation
+Hedera Services will perform a circular scanning of entities, meaning after we reach the last entity in the system, we will go back scanning from the first entity in the system.
+
+When trying to renew an entity:
+1. Calculate the fee to extend the entity's `expirationTime` for another `autoRenewPeriod`.
+2. If the `autoRenewAccount` of the entity has enough balance to cover this fee:
+  - extend the entity's `expirationTime` for another `autoRenewPeriod`.
+  - otherwise, translate the remaining balance of the `autoRenewAccount` into an extension, preferably proportional to the fee calculated in step 1, then extend accordingly.
+3. If the grace period also passes, delete the entity from the system.
```

Comment:
Would an entity be deleted if it waited for the auto renew to handle it but got expired in the wait period? Does this mean that grace period should at least be as big for the auto renew feature to handle all the entities? If so would the grace period grow/shrink with an increase/decrease of entities?

Reply:
The system is expected to do a full scan of all entities in one day at most. The grace period will be defaulted to 7 days so it should be long enough. Also, in a rare situation where we don't have many activities (transactions) during this 7 days to reach an entity that needs autorenewal asap, the system should always try to renew it first before deleting it from the system.

Reply:
Makes sense although it would be good to specify it the document.